### ISSUE TITLE
`reportUndefinedVariable` in nested vs. inline context manager

### ISSUE THREAD

**Issue by lu-pl** [microsoft/pyright#11165]
There appears to be a difference in how `pyright` treats nested vs. inline context managers, e.g when inferring undefined variables.

The following correctly reports `error: "value" is not defined (reportUndefinedVariable)` in `pyright 1.1.407` for `value` in `print(value)`:

```python
from contextlib import contextmanager, suppress

@contextmanager
def context():
    raise RuntimeError
    yield 42

with suppress(RuntimeError):
    with context() as value:
        raise RuntimeError

print(value)
```

`__enter__` in `context` raises before returning, `contextlib.suppress` catches the `RuntimeError` and control is transferred to after the `suppress` block before `value` can be bound.

However, for the same definition with an inline context manager, `pyright` reports `0 errors, 0 warnings, 0 informations`: 

```python
from contextlib import contextmanager, suppress

@contextmanager
def context():
    raise RuntimeError
    yield 42

with suppress(RuntimeError), context() as value:
    raise RuntimeError

print(value)
```

**Comment by erictraut** [comment]
For details about how type checkers handle context managers, refer to [this section of the typing spec](https://typing.python.org/en/latest/spec/exceptions.html#context-managers). Pyright's behavior in in compliance with the spec.

**Comment by lu-pl** [comment]
Thank you, the referenced docs section and [`__exit__ in suppress`](https://github.com/python/cpython/blob/0dd83111ab4f8f29daf0de821f82d2208a0f4873/Lib/contextlib.py#L450) are quite insightful!

I am sorry to bother you further, but imo the referenced section of the Python docs does not explain why `pyright` would behave differently in nested vs. inline/compound context managers.
